packages/client/lib/sync/fetcher/accountfetcher.ts

```diff
@@ -0,0 +1,332 @@
+
+  /** The origin to start account fetcher from */
+  first: bigint
+
+  /** Range to eventually fetch */
+  count?: bigint
+
+  /** Destroy fetcher once all tasks are done */
+  destroyWhenDone?: boolean
+}
+
+// root comes from block?
+export type JobTask = {
+  /** The origin to start account fetcher from */
+  first: bigint
+  /** Range to eventually fetch */
+  count: bigint
+}
+
+export class AccountFetcher extends Fetcher<JobTask, AccountData[], AccountData> {
+  /**
+   * The stateRoot for the fetcher which sorts of pin it to a snapshot.
+   * This might eventually be removed as the snapshots are moving and not static
+   */
+  root: Buffer
+
+  /** The origin to start account fetcher from (including), by default starts from 0 (0x0000...) */
+  first: bigint
+
+  /** The range to eventually, by default should be set at BigInt(2) ** BigInt(256) + BigInt(1) - first */
+  count: bigint
+
+  /**
+   * Create new block fetcher
+   */
+  constructor(options: AccountFetcherOptions) {
+    super(options)
+
+    this.root = options.root
+    this.first = options.first
+    this.count = options.count ?? BigInt(2) ** BigInt(256) - this.first
+
+    const fullJob = { task: { first: this.first, count: this.count } } as Job<
+      JobTask,
+      AccountData[],
+      AccountData
+    >
+    const origin = this.getOrigin(fullJob)
+    const limit = this.getLimit(fullJob)
+
+    this.debug(
+      `Account fetcher instantiated root=${short(this.root)} origin=${short(origin)} limit=${short(
+        limit
+      )} destroyWhenDone=${this.destroyWhenDone}`
+    )
+  }
+
+  private async verifyRangeProof(
+    stateRoot: Buffer,
+    origin: Buffer,
+    { accounts, proof }: { accounts: AccountData[]; proof: Buffer[] }
+  ): Promise<void> {
+    this.debug(
+      `verifyRangeProof accounts:${accounts.length} first=${short(accounts[0].hash)} last=${short(
+        accounts[accounts.length - 1].hash
+      )}`
+    )
+
+    for (let i = 0; i < accounts.length - 1; i++) {
+      // ensure the range is monotonically increasing
+      if (accounts[i].hash.compare(accounts[i + 1].hash) === 1) {
+        throw Error(
+          `Account hashes not monotonically increasing: ${i} ${accounts[i].hash} vs ${i + 1} ${
+            accounts[i + 1].hash
+          }`
+        )
+      }
+    }
+
+    const trie = new Trie({ db: new LevelDB() })
+    const keys = accounts.map((acc: any) => acc.hash)
+    const values = accounts.map((acc: any) => accountBodyToRLP(acc.body))
+    // convert the request to the right values
+    await trie.verifyRangeProof(stateRoot, origin, keys[keys.length - 1], keys, values, <any>proof)
+  }
+
+  private isMissingRightRange(
+    limit: Buffer,
+    { accounts, proof: _proof }: { accounts: AccountData[]; proof: Buffer[] }
+  ): boolean {
+    if (accounts.length > 0 && accounts[accounts.length - 1]?.hash.compare(limit) >= 0) {
+      return false
+    } else {
+      // TODO: Check if there is a proof of missing limit in state
```

Comment:
@jochem-brouwer @scorbajio Is there an early check that can be done which can see there are no more accounts between the last and the limit

Reply:
`verifyRangeProof` returns a boolean if there are more items or not. We could maybe early-check `hasRightElement` but I don't know if this really speeds it up?

Reply:
I was originally thinking that if there is an empty response to a request, it mean there are no accounts in a given range. There might also be a cryptographic way to check this.